### PR TITLE
FR-12043 - allow user to fetch entitlements by keys

### PR DESCRIPTION
**Pull request by doregg**
Disclaimer:
We should test somehow the refresh token and switch tenant.

User component code example:

**Component ts file:**
```ts
export class AppComponent implements OnInit, OnDestroy {
  isLoading = true;
  loadingSubscription: Subscription;
  entitlementsSubscription: Subscription;
  isFeature1Ent: boolean | undefined;
  isFeature2Ent: boolean | undefined;

  constructor(private fronteggAppService: FronteggAppService) {
    this.entitlementsSubscription = this.fronteggAppService.entitlements$(['feature-1', 'feature-2'], {
      next: (
        [{ isEntitled: isFeature1Ent }, { isEntitled: isFeature2Ent }]: Entitlements
      ) => {
        this.isFeature1Ent = isFeature1Ent;
        this.isFeature2Ent = isFeature2Ent;
      }
    });
  }

  ngOnDestroy(): void {
    this.loadingSubscription.unsubscribe();
    this.entitlementsSubscription.unsubscribe();
  }
}
```

Component html file:
```html
<div *ngIf="!isLoading">
    <router-outlet></router-outlet>
    <div *authorizedContent="['Admin']">Authorized content section for Admin's (example)</div>
    <div>Feature 1: {{isFeature1Ent ? 'on' : 'off'}}</div>
    <div>Feature 2: {{isFeature2Ent ? 'on' : 'off'}}</div>
</div>
```

### CODE REVIEW

File: projects/frontegg-app/src/lib/frontegg-app.service.ts [hunk 1]
```diff
@@ -178,4 +183,12 @@ export class FronteggAppService {
   public hideCheckoutDialog(): void {
     this.fronteggApp?.hideCheckoutDialog();
   }
+
+  /**
+    @param keys The requested entitlement keys
+    @returns Entitlements contain true/false for every key (state of is key entitled)
+  */
+  public getEntitlements(keys: string[]): Entitlements {
```
Comment (yuvalotem1) [inline projects/frontegg-app/src/lib/frontegg-app.service.ts:191]: maybe more declarative name then keys?

File: projects/frontegg-app/src/lib/frontegg-app.service.ts [hunk 2]
```diff
@@ -139,6 +143,7 @@ export class FronteggAppService {
 
     this.stateSubject.next(fronteggStore);
     this.authStateSubject.next(fronteggStore.auth);
+    this.entitlementsStateSubject.next(fronteggStore.auth.entitlementsState?.entitlements || null);
```
Comment (yuvalotem1) [inline projects/frontegg-app/src/lib/frontegg-app.service.ts:146]: what would happen if you provide null?

File: projects/frontegg-app/src/lib/frontegg-app.service.ts [hunk 3]
```diff
@@ -29,6 +29,7 @@ export class FronteggAppService {
     isLoading: true,
     isAuthenticated: false,
   } as FronteggState['auth']);
+  private entitlementsStateSubject = new BehaviorSubject<any>(null);
```
Comment (AtaliaRefua) [inline projects/frontegg-app/src/lib/frontegg-app.service.ts:32]: why any?

Comment (yuvalotem1) [reply projects/frontegg-app/src/lib/frontegg-app.service.ts:32]: Entitlements | null

File: projects/frontegg-app/src/lib/frontegg-app.service.ts [hunk 4]
```diff
@@ -63,6 +64,32 @@ export class FronteggAppService {
     return this.isLoadingSubject.asObservable();
   };
 
+  /**
+   * @param keys Entitlements keys
+   * @param observer For receiving the keys data results
+   * @returns a subscription to be able to unsubscribe
+   */
+  public entitlements$(keys: string[], observer: PartialObserver<Entitlements>): Subscription {
```
Comment (yuvalotem1) [inline projects/frontegg-app/src/lib/frontegg-app.service.ts:72]: how would it be used by the our users?

File: projects/frontegg-app/src/lib/frontegg-app.service.ts [hunk 5]
```diff
@@ -63,6 +64,32 @@ export class FronteggAppService {
...
+    // used for computing the entitlements result because we don't return the state itself
+    const entitlementsSubject = new BehaviorSubject<Entitlements>([]);
+    
+    const stateSubscription = this.entitlementsStateSubject.subscribe(entitlements => {
+      entitlementsSubject.next(getEntitlements(entitlements ?? {}, keys));
+    });
+    
+    // subscribing the consumer observer
+    const entitlementsResultSubscription = entitlementsSubject.asObservable().subscribe(observer)
```
Comment (yuvalotem1) [inline projects/frontegg-app/src/lib/frontegg-app.service.ts:81]: Why not saving entitlementsResultSubscription as part of the class? and then could be access from multiple places?

File: projects/frontegg-app/src/lib/frontegg-app.service.ts (hunk 3 above)
Comment (doregg) [reply projects/frontegg-app/src/lib/frontegg-app.service.ts:32]: it's not entitlements.. it's a type I have inside the internal logic.. I'm thinking how to handle it :/ 
lemme know if you have an idea

File: projects/frontegg-app/src/lib/frontegg-app.service.ts (hunk 1 above)
Comment (doregg) [reply projects/frontegg-app/src/lib/frontegg-app.service.ts:214]: this is the terminology we use also in the BE. I wanna be aligned

**Review by yuvalotem1 — APPROVED** [review]

File: projects/frontegg-app/src/lib/frontegg-app.service.ts (hunk 5 above)
Comment (doregg) [reply projects/frontegg-app/src/lib/frontegg-app.service.ts:81]: cause we use different keys for every func call

File: projects/frontegg-app/src/lib/frontegg-app.service.ts (hunk 4 above)
Comment (doregg) [reply projects/frontegg-app/src/lib/frontegg-app.service.ts:72]: description

**Review by rotemzif1 — APPROVED** [review]

**Pull request merged**